client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/object/ClassToSchema.java

```diff
@@ -64,38 +66,89 @@ public static void registerSchemaIfNeeded(Class<?> valueType) {
         }
 
         String name = valueType.getName();
-        String schema = ClassToSchema.getOrDeriveSchema(valueType);
-        UnitsInfoRecorder.registerNewParsedDto(name, schema);
-        ExecutionTracer.addParsedDtoName(name);
+        if (!UnitsInfoRecorder.isDtoSchemaRegister(name)){
+            List<Class<?>> embedded = new ArrayList<>();
+            String schema = ClassToSchema.getOrDeriveSchema(valueType, embedded);
+            UnitsInfoRecorder.registerNewParsedDto(name, schema);
+            ExecutionTracer.addParsedDtoName(name);
+            if (!embedded.isEmpty()){
+                embedded.forEach(ClassToSchema::registerSchemaIfNeeded);
+            }
 
+        }
+    }
+    /**
+     *
+     * @return a schema representation of the class along with schemas of its ref classes in the form
+     *      "name: { "type name": "schema", "type name": "schema", .... }"
+     */
+    public static String getOrDeriveSchemaWithItsRef(Class<?> klass){
+        StringBuilder sb = new StringBuilder();
+        List<Class<?>> nested = new ArrayList<>();
+        String schema = ClassToSchema.getOrDeriveSchemaAndEmbeddedClasses(klass, nested);
+        sb.append("{");
+        sb.append(schema);
+
+        nested.forEach(s->
+                sb.append(",").append(ClassToSchema.getOrDeriveNonNestedSchema(s)));
+
+        sb.append("}");
+        return named(klass.getName(), sb.toString());
     }
-
 
     /**
+     * @param nested is a list of nested classes
      * @return a schema representation of the class in the form "name: {...}", ie
      * like a field entry in an OpenAPI object definition
      */
-    public static String getOrDeriveSchema(Class<?> klass) {
-        return getOrDeriveSchema(klass.getName(), klass);
+    public static String getOrDeriveSchema(Class<?> klass, List<Class<?>> nested) {
+        return getOrDeriveSchema(klass.getName(), klass, false, nested);
     }
 
-    public static String getOrDeriveSchema(String name, Type type) {
+    /**
+     *
+     * @return a schema representation of the class in the form "name: {...}"
+     */
+    public static String getOrDeriveNonNestedSchema(Class<?> klass) {
+        return getOrDeriveSchema(klass.getName(), klass, false, Collections.emptyList());
+    }
 
-        if (cache.containsKey(type)) {
+    private static String getOrDeriveSchema(String name, Type type, Boolean useRefObject, List<Class<?>> nested) {
+
+        if (cache.containsKey(type) && !useRefObject) {
             return named(name, cache.get(type));
```

Comment:
on `core` it is no problem, but on driver side need to be very carefull of performance, especially for string concatenations. might want to add cache to avoid re-computing `named()` each time

Reply:
here, `cache` only saves schema with its properties.

if the `useRefObject` is true, it means that a schema for the object  will be a reference schema (such as. "$ref":"#/components/schemas/org.evomaster.client.java.instrumentation.object.dtos.CycleDtoA").

Reply:
hi. maybe easier to discuss what I meant in a call. remind me tomorrow morning

Reply:
hi, thanks. update it a bit.